ledger/complete/wal/checkpoint_v6_leaf_reader.go

```diff
@@ -0,0 +1,106 @@
+package wal
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/hash"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/flattener"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/node"
+	"github.com/rs/zerolog"
+)
+
+type LeafNode struct {
+	Hash    hash.Hash
+	Path    ledger.Path
+	Payload *ledger.Payload
+}
+
+type LeafNodeResult struct {
+	LeafNode *LeafNode
+	Err      error
+}
+
+func nodeToLeaf(leaf *node.Node) *LeafNode {
+	return &LeafNode{
+		Hash:    leaf.Hash(),
+		Path:    *leaf.Path(),
+		Payload: leaf.Payload(),
+	}
+}
+
+func OpenAndReadLeafNodesFromCheckpointV6(dir string, fileName string, logger *zerolog.Logger) (
+	allLeafNodesCh <-chan LeafNodeResult, errToReturn error) {
+
+	filepath := filePathCheckpointHeader(dir, fileName)
+
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("could not open file %v: %w", filepath, err)
+	}
+	defer func(file *os.File) {
+		errToReturn = closeAndMergeError(file, errToReturn)
+	}(f)
+
+	subtrieChecksums, _, err := readCheckpointHeader(filepath, logger)
+	if err != nil {
+		return nil, fmt.Errorf("could not read header: %w", err)
+	}
+
+	// ensure all checkpoint part file exists, might return os.ErrNotExist error
+	// if a file is missing
+	err = allPartFileExist(dir, fileName, len(subtrieChecksums))
+	if err != nil {
+		return nil, fmt.Errorf("fail to check all checkpoint part file exist: %w", err)
+	}
+
+	bufSize := 1000
+	leafNodesCh := make(chan LeafNodeResult, bufSize)
+	allLeafNodesCh = leafNodesCh
+	defer func() {
+		close(leafNodesCh)
+	}()
+
+	// push leaf nodes to allLeafNodesCh
+	for i, checksum := range subtrieChecksums {
+		readCheckpointSubTrieLeafNodes(leafNodesCh, dir, fileName, i, checksum, logger)
+	}
+
+	return allLeafNodesCh, nil
+}
+
+func readCheckpointSubTrieLeafNodes(leafNodesCh chan<- LeafNodeResult, dir string, fileName string, index int, checksum uint32, logger *zerolog.Logger) {
+	err := processCheckpointSubTrie(dir, fileName, index, checksum, logger,
+		func(reader *Crc32Reader, scratch []byte, nodesCount uint64) error {
```

Comment:
the two inline processNode functions (this in file and in the next file) are nearly identical.  refactor them to share the same read logic plz

Reply:
I'm not sure how to share the same read logic, note, the ReadNode func also takes a function, and there it returns a list of nodes, and here, it returns a channel of leaf nodes.

There are different use cases, there, we need to maintain a `nodes` list, which has to be created outside of the process function and to be returned the `readCheckpointSubTrie`, meaning the `ReadNode` needs to update the `nodes` list which is outside of the process function.
Whereas here, we don't need to update that nodes list, which takes a lot of memory, but just have to push the node being read to the channel.

Given the difference, it's easier to duplicate some code IMO, do I miss anything?

Reply:
you can add a helper function which returns the list of read node?

leaf reader can iterate on this list, check for leaf node, and send to channel

regular read can just use the list directly

am i missing something?

Reply:
>you can add a helper function which returns the list of read node?

Then the helper function will have to create the list of read nodes internally, which would use up lots of memory. 

Whereas the current implementation of `readCheckpointSubTrieLeafNodes`, as soon as it read the node, we will check if it's a leaf node, if it is not a leaf node, then we throw it away. If it is a leaf node, then we push it to the channel, if the channel is full because all workers are busy processing the leaf nodes, then the pushing will be blocked, which is good, because we don't want to read too many nodes in memory, which will increase memory usage. We only want to read enough nodes to feed the workers, and wait until the next worker to be available.

Reply:
return an iterator instead of a list?

Reply:
I'm not sure which function you are talking about.

Do you mean `readCheckpointSubTrieLeafNodes ` to return an iterator instead of taking a channel?

I tried to use iterator, but I found the code change to the `readCheckpointSubTrie` function would too big and become a refactor, which would be much harder to review. 

In the end, I switched to using channel, which allows much less code changes, and easier to review. 

And even if I refactor into iterator, I think I still can't reuse more code, because the iterator won't know whether I need to build a list of nodes or not. The iterator still has to take a function to be passed to `flattener.ReadNode`.